master/src/main/scala/org/apache/celeborn/service/deploy/master/quota/QuotaManager.scala

```diff
@@ -16,96 +16,393 @@
 import org.apache.celeborn.common.identity.UserIdentifier
 import org.apache.celeborn.common.internal.Logging
-import org.apache.celeborn.common.quota.{Quota, ResourceConsumption}
-import org.apache.celeborn.common.util.Utils
+import org.apache.celeborn.common.metrics.source.ResourceConsumptionSource
+import org.apache.celeborn.common.metrics.source.ResourceConsumptionSource._
+import org.apache.celeborn.common.protocol.message.ControlMessages.CheckQuotaResponse
+import org.apache.celeborn.common.quota.{ResourceConsumption, StorageQuota}
+import org.apache.celeborn.common.util.{JavaUtils, ThreadUtils, Utils}
 import org.apache.celeborn.server.common.service.config.ConfigService
+import org.apache.celeborn.service.deploy.master.MasterSource
+import org.apache.celeborn.service.deploy.master.MasterSource.UPDATE_RESOURCE_CONSUMPTION_TIME
+import org.apache.celeborn.service.deploy.master.clustermeta.AbstractMetaManager
+import org.apache.celeborn.service.deploy.master.quota.QuotaStatus._
+
+class QuotaManager(
+    statusSystem: AbstractMetaManager,
+    masterSource: MasterSource,
+    resourceConsumptionSource: ResourceConsumptionSource,
+    celebornConf: CelebornConf,
+    configService: ConfigService) extends Logging {
 
-class QuotaManager(celebornConf: CelebornConf, configService: ConfigService) extends Logging {
-  private val DEFAULT_QUOTA = Quota(
-    celebornConf.get(CelebornConf.QUOTA_DISK_BYTES_WRITTEN),
-    celebornConf.get(CelebornConf.QUOTA_DISK_FILE_COUNT),
-    celebornConf.get(CelebornConf.QUOTA_HDFS_BYTES_WRITTEN),
-    celebornConf.get(CelebornConf.QUOTA_HDFS_FILE_COUNT))
-  def getQuota(userIdentifier: UserIdentifier): Quota = {
-    if (configService != null) {
-      val config =
-        configService.getTenantUserConfigFromCache(userIdentifier.tenantId, userIdentifier.name)
-      config.getQuota
+  val userQuotaStatus: JMap[UserIdentifier, QuotaStatus] = JavaUtils.newConcurrentHashMap()
+  val tenantQuotaStatus: JMap[String, QuotaStatus] = JavaUtils.newConcurrentHashMap()
+  val resourceConsumptionMetricsEnabled = celebornConf.masterResourceConsumptionMetricsEnabled
+  @volatile
+  var clusterQuotaStatus: QuotaStatus = new QuotaStatus()
+  val appQuotaStatus: JMap[String, QuotaStatus] = JavaUtils.newConcurrentHashMap()
+  private val quotaChecker =
+    ThreadUtils.newDaemonSingleThreadScheduledExecutor("master-quota-checker")
+  quotaChecker.scheduleWithFixedDelay(
+    new Runnable {
+      override def run(): Unit = {
+        try {
+          updateResourceConsumption()
+        } catch {
+          case t: Throwable => logError("Update user resource consumption failed.", t)
+        }
+      }
+    },
+    0L,
+    celebornConf.masterResourceConsumptionInterval,
+    TimeUnit.MILLISECONDS)
+
+  def handleAppLost(appId: String): Unit = {
+    appQuotaStatus.remove(appId)
+  }
+
+  def checkUserQuotaStatus(user: UserIdentifier): CheckQuotaResponse = {
+    val quotaEnabled =
```

Comment:
I think it's acceptable to verify `_.getSystemConfigFromCache().quotaEnabledForTenant/User/Cluster ` to determine if quota checking is enabled at the system level. Supporting configuration at the tenant or user level adds unnecessary complexity.

Reply:
Should we keep quotaEnabled?

Reply:
quotaEnabled can be the global configuration to decide whether enable `QuotaManager`, seems we can also support dynamic configuration for that.